pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,600 @@
+	m.numAllocated--
+	m.removed++
+
+	if m.removed%mapReInitializeThreshold == 0 {
+		n := make(map[string]uint32)
+		for k, v := range m.usedBy {
+			n[k] = v
+		}
+		m.usedBy = n
+		m.removed = 0
+	}
+	m.used.clear(v / userNsLength)
+}
+
+func (m *usernsManager) parseUserNsFileAndRecord(pod types.UID, content []byte) (userNs userNamespace, err error) {
+	if err = json.Unmarshal([]byte(content), &userNs); err != nil {
+		err = fmt.Errorf("can't parse file: %w", err)
+		return
+	}
+
+	if len(userNs.UIDMappings) != 1 {
+		err = fmt.Errorf("invalid user namespace configuration: no more than one mapping allowed.")
+		return
+	}
+
+	if len(userNs.UIDMappings) != len(userNs.GIDMappings) {
+		err = fmt.Errorf("invalid user namespace configuration: GID and UID mappings should be identical.")
+		return
+	}
+
+	if userNs.UIDMappings[0] != userNs.GIDMappings[0] {
+		err = fmt.Errorf("invalid user namespace configuration: GID and UID mapping should be identical")
+		return
+	}
+
+	// We don't produce configs without root mapped and some runtimes assume it is mapped.
+	// Validate the file has something we produced and can digest.
+	if userNs.UIDMappings[0].ContainerId != 0 {
+		err = fmt.Errorf("invalid user namespace configuration: UID 0 must be mapped")
+		return
+	}
+
+	if userNs.GIDMappings[0].ContainerId != 0 {
+		err = fmt.Errorf("invalid user namespace configuration: GID 0 must be mapped")
+		return
+	}
+
+	hostId := userNs.UIDMappings[0].HostId
+	length := userNs.UIDMappings[0].Length
+
+	err = m.record(pod, hostId, length)
+	return
+}
+
+func (m *usernsManager) createUserNs(pod *v1.Pod) (userNs userNamespace, err error) {
+	firstID, length, err := m.allocateOne(string(pod.UID))
+	if err != nil {
+		return
+	}
+
+	defer func() {
+		if err != nil {
+			m.Release(pod.UID)
+		}
+	}()
+
+	userNs = userNamespace{
+		UIDMappings: []idMapping{
+			{
+				ContainerId: 0,
+				HostId:      firstID,
+				Length:      length,
+			},
+		},
+		GIDMappings: []idMapping{
+			{
+				ContainerId: 0,
+				HostId:      firstID,
+				Length:      length,
+			},
+		},
+	}
+
+	return userNs, m.writeMappingsToFile(pod.UID, userNs)
+}
+
+// GetUserNamespaceMappings returns the configuration for the sandbox user namespace
+func (m *usernsManager) GetUserNamespaceMappings(pod *v1.Pod) (*runtimeapi.UserNamespace, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		return nil, nil
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	if pod.Spec.HostUsers == nil || *pod.Spec.HostUsers == true {
+		return &runtimeapi.UserNamespace{
+			Mode: runtimeapi.NamespaceMode_NODE,
+		}, nil
+	}
+
+	content, err := m.readMappingsFromFile(pod.UID)
+	if err != nil && err != utilstore.ErrKeyNotFound {
+		return nil, err
+	}
+
+	var userNs userNamespace
+	if string(content) != "" {
+		userNs, err = m.parseUserNsFileAndRecord(pod.UID, content)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		userNs, err = m.createUserNs(pod)
```

Comment:
is this the only section that needs to be protected via the lock as its the only section that is mutating state on disk?

Reply:
No, all "public" functions should be protected. The file on disk is not the only thing, but we have maps for example: maps can't be written to simultaneously (that creates a panic in go), nor should be edited while other go-routine is reading them.

So, we need the lock also on Release(), CleanupOrphanedPodUsernsAllocations(), that is the functions we take the lock on :)

Reply:
Holding a lock while performing I/O seems like a bad idea.